### ISSUE TITLE
requirements on setup.py

### ISSUE THREAD

**Issue by avaldebe** [metno/pyaerocom#395]
plan to submit a PR later this week with addressing the following

- [x] update classifiers
- [x] add minimum python version 
- [x] add dependencies
- [x] use the correct python version on CI

**Comment by jgliss** [comment]
@avaldebe not sure what is the best way to handle dependencies but I would prefer to have a single location where we have all requirements. Currently, this is the [pyaerocom_env.yml](https://github.com/metno/pyaerocom/blob/master/pyaerocom_env.yml) which is also used in the [CI.yml](https://github.com/metno/pyaerocom/blob/master/.github/workflows/CI.yml) workflow file. Perhaps, *setup.py* could just import requirements from *pyaerocom_env.yml*?

**Comment by avaldebe** [comment]
@jgliss 

It is not straight forward as the `iris` conda package correspond to the `scitools-iris` [python package](https://pypi.org/project/scitools-iris/) , and I'm not sure if the python package version matches the conda package version.

In any case, installing the requirements on a Python 3.6 virtual environment has been challenging.
`scitools-iris` requires `pyke`, which needs to be installed as `scitools-pyke`...

Also, `scitools-iris` recently dumped support for Python 3.6
Maybe the minimum versions of `scitools-iris` and Python should be bumped.

How about using a build tool that uses `pyproject.toml`?
I quite like poetry, because of its dependency resolver.

**Comment by jgliss** [comment]
@avaldebe I understand. I would quite like to keep conda as the main way of installing pyaerocom (preferably, users should use the latest version via *conda-forge*). That way, dependencies are also resolved nicely. I am not familiar with these other build tools, like `poetry`, but if you have a nice alternative solution to the conda option, of course we can add that! And I would say we drop support for P3.6 in any case. What do you think?

**Comment by avaldebe** [comment]
> @avaldebe I understand. I would quite like to keep conda as the main way of installing pyaerocom (preferably, users should use the latest version via _conda-forge_). That way, dependencies are also resolved nicely. 

OK, but this would mean duplicating the dependencies. That way we can be explicit with the conda/python package names and versions.

> I am not familiar with these other build tools, like `poetry`, but if you have a nice alternative solution to the conda option, of course we can add that! And I would say we drop support for P3.6 in any case. What do you think?

Using [poetry][] would replace `setup.py` by `pyproject.toml`, but that requires that poetry is installed on your development machine. My preferred way to install python tools is with [pipx][]. I can provide detailed instructions if you decide to swap to poetry

[poetry]: https://python-poetry.org/docs/basic-usage/
[pipx]: https://pypa.github.io/pipx/

**Comment by jgliss** [comment]
Thanks, I will need to check a bit more on poetry, but for now I think we should stick to conda + setup.py, which does management of virtualenvs well, as well as solving requirements. Also, it seems to me that still, most of the python world sticks to more classic approaches, recommending to install via conda, pip, etc, e.g [numpy](https://numpy.org/install/). Also not sure how well poetry is able to deal with binary dependencies, etc.

And I quite like to work in development mode, i.e. being able to do `python setup.py develop` to install my dev repo in any conda env, without need of copying the file base. If I understand correctly, this may not be as trivial with poetry. 

On the other hand, having one place to manage dependencies, installation as well as deployment seems rather tempting to me, so we should keep an eye on poetry! Do you know if there is any symbiosis planned between conda and poetry? I.e., that a user that works with conda could install a `poetry` package easily, and cross resolve dependencies?

**Comment by avaldebe** [comment]
> Thanks, I will need to check a bit more on poetry, but for now I think we should stick to conda + setup.py, which does management of virtualenvs well, as well as solving requirements.
> Also, it seems to me that still, most of the python world sticks to more classic approaches, recommending to install via conda, pip, etc, e.g [numpy](https://numpy.org/install/). Also not sure how well poetry is able to deal with binary dependencies, etc.

A package build with `poetry` will be installable with pip. It replaces `setuptools` when building the wheels and tar archives.
For local development/testing, it will solve dependencies and install the required packages on a virtual environment.
The installation of the resolved dependencies is done with by pip (or its internals).

In any case, I'll use poetry to generate my virtual environment and ignore the `pyproject.toml` and `poetry.lock` files.

BTW, is `nbsphinx` needed for the operation of the package?
Can it be set as an extra? e.g. `pip install pyaerocom[docs]`
Or is a development/test only dependency?

Here are its dependency tree, with the dependencies resolved by poetry

```
$ poetry show nbsphinx --tree
nbsphinx 0.8.5 Jupyter Notebook Tools for Sphinx
├── docutils *
├── jinja2 >=2.11,<3.0
│   └── markupsafe >=0.23 
├── nbconvert !=5.4
│   ├── bleach * 
│   │   ├── packaging * 
│   │   │   └── pyparsing >=2.0.2 
│   │   ├── six >=1.9.0 
│   │   └── webencodings * 
│   ├── defusedxml * 
│   ├── entrypoints >=0.2.2 
│   ├── jinja2 >=2.4 
│   │   └── markupsafe >=0.23 
│   ├── jupyter-core * 
│   │   ├── pywin32 >=1.0 
│   │   └── traitlets * 
│   │       ├── decorator * 
│   │       ├── ipython-genutils * 
│   │       └── six * (circular dependency aborted here)
│   ├── jupyterlab-pygments * 
│   │   └── pygments >=2.4.1,<3 
│   ├── mistune >=0.8.1,<2 
│   ├── nbclient >=0.5.0,<0.6.0 
│   │   ├── async-generator * 
│   │   ├── jupyter-client >=6.1.5 
│   │   │   ├── jupyter-core >=4.6.0 (circular dependency aborted here)
│   │   │   ├── nest-asyncio >=1.5 
│   │   │   ├── python-dateutil >=2.1 
│   │   │   │   └── six >=1.5 (circular dependency aborted here)
│   │   │   ├── pyzmq >=13 
│   │   │   │   ├── cffi * 
│   │   │   │   │   └── pycparser * 
│   │   │   │   └── py * 
│   │   │   ├── tornado >=4.1 
│   │   │   └── traitlets * (circular dependency aborted here)
│   │   ├── nbformat >=5.0 
│   │   │   ├── ipython-genutils * (circular dependency aborted here)
│   │   │   ├── jsonschema >=2.4,<2.5.0 || >2.5.0 
│   │   │   │   ├── attrs >=17.4.0 
│   │   │   │   ├── importlib-metadata * 
│   │   │   │   │   ├── typing-extensions >=3.6.4 
│   │   │   │   │   └── zipp >=0.5 
│   │   │   │   ├── pyrsistent >=0.14.0 
│   │   │   │   └── six >=1.11.0 (circular dependency aborted here)
│   │   │   ├── jupyter-core * (circular dependency aborted here)
│   │   │   └── traitlets >=4.1 (circular dependency aborted here)
│   │   ├── nest-asyncio * (circular dependency aborted here)
│   │   └── traitlets >=4.2 (circular dependency aborted here)
│   ├── nbformat >=4.4 (circular dependency aborted here)
│   ├── pandocfilters >=1.4.1 
│   ├── pygments >=2.4.1 (circular dependency aborted here)
│   ├── testpath * 
│   └── traitlets >=4.2 (circular dependency aborted here)
├── nbformat *
│   ├── ipython-genutils * 
│   ├── jsonschema >=2.4,<2.5.0 || >2.5.0 
│   │   ├── attrs >=17.4.0 
│   │   ├── importlib-metadata * 
│   │   │   ├── typing-extensions >=3.6.4 
│   │   │   └── zipp >=0.5 
│   │   ├── pyrsistent >=0.14.0 
│   │   └── six >=1.11.0 
│   ├── jupyter-core * 
│   │   ├── pywin32 >=1.0 
│   │   └── traitlets * 
│   │       ├── decorator * 
│   │       ├── ipython-genutils * (circular dependency aborted here)
│   │       └── six * (circular dependency aborted here)
│   └── traitlets >=4.1 (circular dependency aborted here)
├── sphinx >=1.8
│   ├── alabaster >=0.7,<0.8 
│   ├── babel >=1.3 
│   │   └── pytz >=2015.7 
│   ├── colorama >=0.3.5 
│   ├── docutils >=0.14,<0.18 
│   ├── imagesize * 
│   ├── jinja2 >=2.3 
│   │   └── markupsafe >=0.23 
│   ├── packaging * 
│   │   └── pyparsing >=2.0.2 
│   ├── pygments >=2.0 
│   ├── requests >=2.5.0 
│   │   ├── certifi >=2017.4.17 
│   │   ├── chardet >=3.0.2,<5 
│   │   ├── idna >=2.5,<3 
│   │   └── urllib3 >=1.21.1,<1.27 
│   ├── snowballstemmer >=1.1 
│   ├── sphinxcontrib-applehelp * 
│   ├── sphinxcontrib-devhelp * 
│   ├── sphinxcontrib-htmlhelp * 
│   ├── sphinxcontrib-jsmath * 
│   ├── sphinxcontrib-qthelp * 
│   └── sphinxcontrib-serializinghtml * 
└── traitlets *
    ├── decorator * 
    ├── ipython-genutils * 
    └── six * 
```

**Comment by avaldebe** [comment]
where the `LatLon23` module comes from?

```
$ poetry run pytest
[...]
.venv/lib/python3.6/site-packages/geonum/topodata.py:21: in <module>
    from LatLon23 import LatLon
E   ModuleNotFoundError: No module named 'LatLon23'

**Comment by avaldebe** [comment]
I see, it is not listed on the `geonum` [setup.py][] requirements but on [requirements.txt][]

[setup.py]: https://github.com/jgliss/geonum/blob/master/setup.py#L30
[requirements.txt]: https://github.com/jgliss/geonum/blob/master/requirements.txt

**Comment by avaldebe** [comment]
I added `LatLon23` and `SRTM.py` to the requirements, so I do not longer get `ModuleNotFoundError` errors.
However pytest is not happy....

```
$ poetry run pytest
[...]
FAILED pyaerocom/io/test/test_ebas_file_index.py::test_EbasFileIndex_ALL_STATION_NAMES - TypeError: argument 1 must be str, not PosixPath
FAILED pyaerocom/io/test/test_ebas_file_index.py::test_EbasFileIndex_ALL_STATION_CODES - TypeError: argument 1 must be str, not PosixPath
FAILED pyaerocom/io/test/test_ebas_file_index.py::test_EbasFileIndex_ALL_STATISTICS_PARAMS - TypeError: argument 1 must be str, not PosixPath
FAILED pyaerocom/io/test/test_ebas_file_index.py::test_EbasFileIndex_ALL_VARIABLES - TypeError: argument 1 must be str, not PosixPath
FAILED pyaerocom/io/test/test_ebas_file_index.py::test_EbasFileIndex_ALL_MATRICES - TypeError: argument 1 must be str, not PosixPath
FAILED pyaerocom/io/test/test_ebas_file_index.py::test_EbasFileIndex_ALL_INSTRUMENTS - TypeError: argument 1 must be str, not PosixPath
FAILED pyaerocom/io/test/test_ebas_file_index.py::test_EbasFileIndex_get_table_names - TypeError: argument 1 must be str, not PosixPath
FAILED pyaerocom/io/test/test_ebas_file_index.py::test_EbasFileIndex_get_column_names[station-names0] - TypeError: argument 1 must be str, not PosixPath
FAILED pyaerocom/io/test/test_ebas_file_index.py::test_EbasFileIndex_get_column_names[variable-names1] - TypeError: argument 1 must be str, not PosixPath
```

**Comment by avaldebe** [comment]
> However pytest is not happy....
> 
> ```
> $ poetry run pytest
> [...]
> FAILED pyaerocom/io/test/test_ebas_file_index.py::test_EbasFileIndex_ALL_STATION_NAMES - TypeError: argument 1 must be str, not PosixPath
> ```

`EbasSQLRequest.database` is a `Path`, but on Python 3.6 [sqlite3.connect](https://docs.python.org/3/library/sqlite3.html#sqlite3.connect) needs a str.
Path-like support was added on Python 3.7. Not sure how Python 3.6 passes CI.

In any case, I'll commit a simple patch that ensures that `sqlite3.connect` gets a str.

**Comment by jgliss** [comment]
Hmmm, not sure what is going wrong there but it seems to work in the CI checks in your PR #396. Maybe some other version of sqlite3 when using poetry? One that does not like `pathlib.Path` instances when calling `sqlite3.connect(path)`?

**Comment by jgliss** [comment]
Ha!

**Comment by jgliss** [comment]
Thanks! Should be in `conftest.py`

**Comment by jgliss** [comment]
EBAS_SQLite_DB = EBAS_FILEDIR.parent.joinpath('ebas_file_index.sqlite3')

assert EBAS_SQLite_DB.exists()

**Comment by avaldebe** [comment]
eda4d89c patches the `sqlite3.connect` calls, so it does not mater that `EbasSQLRequest._database` is a `str` or `Path`. The `None` case will raise an exception on `EbasSQLRequest.database`, so it is not a problem.

First I set `EbasSQLRequest.database` to return `str(db)`, but without type hints it seemed more of a hack.
I quite like `Path` objects, and without type hints I decided that `sqlite3.connect(str(self.database))` was a cleaner solution.

In any case, all the test pass on my local setup.

**Comment by avaldebe** [comment]
> Hmmm, not sure what is going wrong there but it seems to work in the CI checks in your PR #396. Maybe some other version of sqlite3 when using poetry? One that does not like `pathlib.Path` instances when calling `sqlite3.connect(path)`?

As far as I can tell the, `python-version` on the CI matrix is the system python, but that is never use because the test run on a conda environment. I added a `python --version` to Conda Info section of `CI.yml` so we'll know soon if I'm correct.

**Comment by avaldebe** [comment]
Looks the [build (3.6)](https://github.com/metno/pyaerocom/pull/396/checks?check_run_id=2737695858) tests are run wuth Python 3.8

```
conda 4.9.2
/usr/share/miniconda/bin/python
Python 3.8.5

**Comment by avaldebe** [comment]
The `s-weigand/setup-conda` action has a [matrix-testing example](https://github.com/s-weigand/setup-conda#matrix-testing).
Based on that the `CI.yaml` should be something like

```YAML
jobs:
  build:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        python-version: [3.6, 3.7, 3.8, 3.9]

    steps:
    - uses: s-weigand/setup-conda@v1.0.5
      with:
        python-version: ${{ matrix.python-version }}
    - name: Conda info
      run: |
        conda --version
        which python
[...]
```

**Comment by avaldebe** [comment]
[build (3.6)](https://github.com/metno/pyaerocom/pull/396/checks?check_run_id=2737929153) tests are now run with Python 3.6

```
conda 4.10.1
/usr/share/miniconda/bin/python
Python 3.6.13 :: Anaconda, Inc.
```

**Comment by avaldebe** [comment]
@jgliss ?

> BTW, is `nbsphinx` needed for the operation of the package?
> Can it be set as an extra? e.g. `pip install pyaerocom[docs]`
> Or is a development/test only dependency?

**Comment by jgliss** [comment]
@avaldebe I need to check that but it was a quickfix so the docs would build on RTD. I think it was due to the fact that RTD ignores requirements.txt files if a conda env file is used

**Comment by avaldebe** [comment]
> [build (3.6)](https://github.com/metno/pyaerocom/pull/396/checks?check_run_id=2737929153) tests are now run with Python 3.6

... but it takes too long update the base environment, so the builds gets cancelled after 6h.
I tried and failed to create a new environment, see 2065d678.
Maybe you have more luck.

> ... need to check that but it was a quickfix so the docs would build on RTD. I think it was due to the fact that RTD ignores requirements.txt files if a conda env file is used

Now that all dependencies are defined on `setup.py`, conda is no longer needed for RTD.
If I define `nbsphinx` as en extra, so the following config `.readthedocs.yml` should work.

```YAML
version: 2

sphinx:
   configuration: docs/conf.py

formats:
   - pdf

build:
  apt_packages:
    - libudunits2-dev
    - libgeos-dev
    - libproj-dev
    - proj-data
    - proj-bin

python:
  version: 3.7
  install:
    - method: pip
      path: .
      extra_requirements:
        - docs
```

I'm not sure if the `build/apt_packages` is needed, but those are the packages I had to install on my local machine for `iris` to work.

**Comment by avaldebe** [comment]
addressed on #396
